loopchain/blockchain/blocks/v0_4/block_verifier.py

```diff
@@ -0,0 +1,211 @@
+
+        if header.height > 0:
+            self.verify_leader_votes(block, prev_block, reps)
+
+        if header.height > 1:
+            prev_next_reps_hash = prev_block.header.revealed_next_reps_hash
+            if prev_next_reps_hash:
+                if prev_next_reps_hash != header.reps_hash:
+                    exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
+                                             f"RepsHash({header.reps_hash}), "
+                                             f"Expected({prev_next_reps_hash}).")
+                    self._handle_exception(exception)
+                prev_reps = reps_getter(prev_block.header.reps_hash)
+            else:
+                prev_reps = reps
+            self.verify_prev_votes(block, prev_reps)
+
+        builder = BlockBuilder.from_new(block, self._tx_versioner)
+        builder.reset_cache()
+        builder.peer_id = block.header.peer_id
+        builder.signature = block.header.signature
+        builder.reps = reps
+
+        for tx in body.transactions.values():
+            builder.transactions[tx.hash] = tx
+
+        invoke_result = None
+        if self.invoke_func:
+            invoke_result = self.verify_invoke(builder, block, prev_block)
+
+        builder.build_transactions_hash()
+        if header.transactions_hash != builder.transactions_hash:
+            exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
+                                     f"TransactionsRootHash({header.transactions_hash.hex()}), "
+                                     f"Expected({builder.transactions_hash.hex()}).")
+            self._handle_exception(exception)
+
+        builder.build_reps_hash()
+        if header.reps_hash != builder.reps_hash:
+            exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
+                                     f"RepRootHash({header.reps_hash.hex()}), "
+                                     f"Expected({builder.reps_hash.hex()}).")
+            self._handle_exception(exception)
+
+        builder.build_prev_votes_hash()
+        if header.prev_votes_hash != builder.prev_votes_hash:
+            exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
+                                     f"PrevVoteRootHash({header.prev_votes_hash.hex()}), "
+                                     f"Expected({builder.prev_votes_hash.hex()}).")
+            self._handle_exception(exception)
+
+        builder.build_hash()
+        if header.hash != builder.hash:
+            exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
+                                     f"Hash({header.hash.hex()}, "
+                                     f"Expected({builder.hash.hex()}), "
+                                     f"header({header}), "
+                                     f"builder({builder.build_block_header_data()}).")
+            self._handle_exception(exception)
+
```

Comment:
`verifiy_generator` is no more needed, right?

Reply:
It would be better to revive with strict `blockchain.get_expected_generator()`.